sycl/source/detail/queue_impl.cpp

```diff
@@ -85,49 +135,55 @@ event queue_impl::memset(const std::shared_ptr<detail::queue_impl> &Self,
   // Emit a begin/end scope for this call
   PrepareNotify.scopedNotify((uint16_t)xpti::trace_point_type_t::task_begin);
 #endif
+
   if (MGraph.lock()) {
     throw sycl::exception(make_error_code(errc::invalid),
                           "The memset feature is not yet available "
                           "for use with the SYCL Graph extension.");
   }
 
-  if (MHasDiscardEventsSupport) {
-    MemoryManager::fill_usm(Ptr, Self, Count, Value,
-                            getOrWaitEvents(DepEvents, MContext), nullptr);
-    return createDiscardedEvent();
-  }
-
-  event ResEvent = prepareSYCLEventAssociatedWithQueue(Self);
+  // We need to submit command and update the last event under same lock if we
+  // have in-order queue.
   {
-    // We need to submit command and update the last event under same lock if we
-    // have in-order queue.
-    auto ScopeLock = isInOrder() ? std::unique_lock<std::mutex>(MLastEventMtx)
-                                 : std::unique_lock<std::mutex>();
-    // If the last submitted command in the in-order queue is host_task then
-    // wait for it before submitting usm command.
-    if (isInOrder() && MLastCGType == CG::CGTYPE::CodeplayHostTask)
-      MLastEvent.wait();
-
-    auto EventImpl = detail::getSyclObjImpl(ResEvent);
-    MemoryManager::fill_usm(Ptr, Self, Count, Value,
-                            getOrWaitEvents(DepEvents, MContext),
-                            &EventImpl->getHandleRef(), EventImpl);
-
-    if (MContext->is_host())
-      return MDiscardEvents ? createDiscardedEvent() : event();
-
+    std::unique_lock<std::mutex> guard(MMutex, std::defer_lock);
+    EventImplPtr ExtraEventToWait = nullptr;
     if (isInOrder()) {
-      MLastEvent = ResEvent;
-      // We don't create a command group for usm commands, so set it to None.
-      // This variable is used to perform explicit dependency management when
-      // required.
-      MLastCGType = CG::CGTYPE::None;
+      guard.lock();
+      ExtraEventToWait = MGraph.expired() ? MLastEventPtr : MGraphLastEventPtr;
```

Comment:
Can we just `push_back` it into the `DepEvents`?

Reply:
This is to avoid copying the vector since it's passed by const reference.

Reply:
Since we have another source of extra events in the form of the `sycl_ext_oneapi_in_order_queue_events` extension now, I ended up reusing the same copied vector that's used for that.